gadopt/gplates/gplates.py

```diff
@@ -113,6 +147,91 @@ def __init__(self, function_space, *args, gplates_connector=None, top_boundary_m
         # Store the GPlates connector
         self.gplates_connector = gplates_connector
 
+        # Store the kwargs in case we need to change solver parameters or surface measure
+        self.kwargs = kwargs
+
+        # Set up the solver for tangential projection
+        self.setup_solver()
+
+    def setup_solver(self):
+        """Set up the linear variational solver for removing radial component."""
+        # Define the solution in the velocity function space
+        # If velocity is discontinuous, we need to use a continuous equivalent
+        if not is_continuous(self):
+            raise ValueError(
+                "GplatesVelocityFunction: Velocity field is discontinuous."
+                "Removing the radial component of discontinuous velocity fields is not supported."
+            )
+        else:
+            V = fd.FunctionSpace(self.function_space().mesh(), self.ufl_element())
+
+        self.tangential_velocity = fd.Function(V, name="tangential_velocity")
+
+        # Normal vector (radial direction for spherical geometry)
+        r = fd.FacetNormal(self.function_space().mesh())
+
+        # Define the test and trial functions for a projection solve
+        phi = fd.TestFunction(V)
+        v = fd.TrialFunction(V)
+
+        # Define the tangential projection: v_tangential = v - (v·r)r
+        tangential_expr = self - fd.inner(self, r) * r
+
+        # Getting ds measure: this can be set by the user through kwargs
+        # We are only looking at the surface measure at the top boundary
+        ds_kwargs = {
+            "domain": self.function_space().mesh(),
+            "degree": 2 * V.ufl_element().degree()[0] + 1
+        }
+
+        # Update the ds_kwargs with the kwargs (only if key already exists in
+        # ds_kwargs, we don't want to add non-relevant keys)
+        ds_kwargs.update({k: v for k, v in self.kwargs.items() if k in ds_kwargs})
+
+        if ds_kwargs["domain"].extruded:
+            self.ds = fd.ds_t(**ds_kwargs)
+        else:
+            ds_kwargs["subdomain_id"] = self.top_boundary_marker
+            self.ds = fd.ds(**ds_kwargs)
+
+        # Setting up a manual projection
+        # Project onto the tangential space: solve for tangential component
+        a = fd.inner(phi, v) * self.ds
+        L = fd.inner(phi, tangential_expr) * self.ds
+
+        # Setting up boundary condition, problem and solver
+        # The field is only meaningful on the boundary, so set zero everywhere else
+        self.interior_null_bc = InteriorBC(V, 0., [self.top_boundary_marker])
+
+        self.problem = fd.LinearVariationalProblem(a, L, self.tangential_velocity,
+                                                   bcs=self.interior_null_bc,
+                                                   constant_jacobian=True)
+        self.solver = fd.LinearVariationalSolver(
+            self.problem,
+            solver_parameters=self.kwargs.get("solver_parameters", self.tangential_project_solver_parameters),
```

Comment:
If you go ahead with my suggestion about the `ds` keyword arguments, then you could also have `solver_parameters` as a keyword argument.